AdvLoggerPkg/Application/DecodeUefiLog/DecodeUefiLog.py

```diff
@@ -598,22 +595,123 @@ def _GetLines(self, lines, LoggerInfo):
+                ThisIsGood = True
+                # We need to verify that from this point on, we have at least
+                # one chunk of eligible entries. If not, we need to move along
+                # the pointer.
+                while CurrentOffset < LoggerInfo["LogBufferSize"] and\
+                        CurrentOffset < StartOffset + self.MAX_MESSAGE_SIZE:
+                    (MessageEntry, _) = self._ReadMessageEntry(LoggerInfo)
+                    CurrentOffset = LoggerInfo["InFile"].tell()
+                    if MessageEntry is None:
+                        ThisIsGood = False
+                        break
+
+                if ThisIsGood:
+                    LoggerInfo["InFile"].seek(StartOffset)
+                    break
+                else:
+                    # We found the first legible line, so we can start from here.
+                    LogStream = LoggerInfo["InFile"].read()
+                    StartOffset = LogStream.find(b'ALM2')
+                    if StartOffset == -1:
+                        StartOffset = LogStream.find(b'ALMS')
+                        if StartOffset == -1:
+                            print("DecodeUefiLog unable to find ALM* signature. Using the beginning as start")
+                            StartOffset = OriginalOffset
+                            break
+                        else:
+                            StartOffset += CurrentOffset
+                    else:
+                        StartOffset += CurrentOffset
+
+            # By the time we are here, we LoggerInfo["InFile"] pointing at the first StartOffset.
+            MessageEntries = []
+            if StartOffset != OriginalOffset:
+                MaximumOffset = LoggerInfo["LogBufferSize"]
+                NeedWrap = True
+            else:
+                # To be honest, there might be some prints that we missed during
+                # querying the logger buffer. We can deal with this nuance later.
+                MaximumOffset = LoggerInfo["LogCurrent"]
+                NeedWrap = False
+
+            # Ok, let's go
+            CurrentOffset = StartOffset
+            while CurrentOffset < MaximumOffset:
+                (MessageEntry, _) = self._ReadMessageEntry(LoggerInfo)
+                CurrentOffset = LoggerInfo["InFile"].tell()
+                if MessageEntry is None:
+                    break
+                MessageEntries.append(MessageEntry)
+
+            # If we need to wrap, we need to start from the beginning.
+            if NeedWrap:
+                CurrentOffset = OriginalOffset
+                MaximumOffset = LoggerInfo["LogCurrent"]
+                LoggerInfo["InFile"].seek(CurrentOffset)
+                while CurrentOffset < MaximumOffset:
+                    (MessageEntry, _) = self._ReadMessageEntry(LoggerInfo)
+                    CurrentOffset = LoggerInfo["InFile"].tell()
+                    if MessageEntry is None:
+                        break

```

Comment:
If we hit this case, this would be an error, right? Because we had a log where we wrapped, but had garbage before LogCurrent, i.e. the most recent message(s) got garbled. We should probably at least raise a warning here, something along the lines of, I tried to parse the log, but the most recent messages have been lost, here is all I could figure out

Reply:
Updated.